Update to support pr 1727 of Neko

timfelle opened the pull request:
Changes tot the time structure in Neko require us to update our interfaces.

The neko work on this started with [PR #1727](https://github.com/ExtremeFLOW/neko/pull/1727), but have been superseeded by [PR #1803](https://github.com/ExtremeFLOW/neko/pull/1803)

timfelle committed:
Feature/MMA (#28)

This PR includes the first iteration of CPU based Method of Moving
Asymptotes.

Remaining tasks:

- [ ] Create interface to align it with use of neko structures.
- [ ] Convert current inline documentation to be doxygen compliant.
- [x] Cleanup redundant code, example call free from init instead of
manually deallocating each array.
- [ ] Add copyright statement.
- [x] Delete test specific code from non test files.
- [ ] Use the neko logging system to add messages to the log.
- [ ] Implement the MMA test, as an actual test unit test.
- [x] Split the test cases to avoid massive comment blocks.

timfelle committed:
MMA/interface (#33)

Split the MMA code to submodules.

timfelle committed:
Refactor code for vector and matrix usage (#34)

This pull request includes several commits that refactor the code to use
vector and matrix data structures for internal MMA parameters.
Additionally, an initializer for JSON files has been added. Work has
also started on supporting other input interfaces. The backend has been
copied to build a vector version, and the progress of compiling the
vector version has been tracked. Arrays have been eliminated, and there
have been minor changes and the deactivation of a broken file.

timfelle committed:
Remove implementation specific imports to implementation (#35)

This pull request removes implementation specific imports to
implementation.

timfelle committed:
Fix issues with Neko tool accessibility and mesh generation (#36)

This pull request addresses two issues. Firstly, it ensures that the
Neko tool is accessible from the examples. Secondly, it runs mesh
generation in the preparation step. The changes include updating the
mesh file path and adding using a Neko tool to generate a mesh box.
Additionally, the initializer in the simcomp module has been moved.

timfelle committed:
Add action for working with flint (#38)

This pull request adds a new action for working with the Flint linter.
The action allows users to execute the Flint linter on a specified
directory, providing options for depth, configuration file, and
outputting the linting score. The action sets up the Python environment,
installs the linter, and runs the linting process. Additionally, the
pull request includes the necessary dependencies for the action.

timfelle committed:
Improvements to the CI (#29)

The goal here is to update the CI.

- [ ] Add Flint support.
- [ ] Add documentation to actions and workflows.
- [ ] Improve reusability and control flow of checks.

timfelle committed:
Disable the formatter (#39)

This pull request disables the formatter by commenting out the
`check_formatting` job in the workflow file.

timfelle committed:
Linting CI improvements (#40)

- Fixes bug in a few linting rules
- Improves structure of linting check
- Ensure suggested changes can be accessed

timfelle committed:
Fix bug in CI (#41)

This pull request fixes a bug in the CI workflow. The bug was causing
the CI workflow to run on all branches, instead of just the `ci/*`
branches. The issue has been resolved by removing the branch restriction
from the workflow configuration.

timfelle committed:
Feature/adjoint (#12)

timfelle committed:
Ensure CUDA_DIR is assigned if asked for. (#42)

Verify device support in Neko after installation

timfelle committed:
Enhancement/control flow pleeease (#37)

here comes the linting....

timfelle committed:
Run compilation tests parallel to linting (#44)

This pull request updates the workflow to run compilation tests in
parallel with linting. The previous workflow only ran linting and
formatting checks. Now, the compilation tests will be executed alongside
the linting process.

timfelle committed:
Feature/pde filters (#46)

Including the PDE filter. I had an email exchange with Niels last night
about filtering of the sensitivities and now they look MUCH better!
Also included some mapping functions and organized things a little bit.

gloopydoop committed:
Fix compilation bugs with missing precision (#47)

This pull request fixes compilation bugs related to missing precision in
the code. The changes include adding compile options for Fortran code,
updating module dependencies, and fixing subroutine arguments. These
changes ensure that the code compiles correctly and avoids
precision-related issues.

timfelle committed:
bugfix (#49)

This shouldn't have been commented out in PR 47 `fix compilation bugs
with missing precision`

timfelle committed:
Align interfaces with changes in Neko (#50)

This pull request aligns the interfaces with the changes made in Neko.
It updates the usage of the `sampler_t` module to the
`output_controller_t` module in multiple files. These changes ensure
that the code is compatible with the latest version of Neko.

timfelle committed:
Clean up redundant variables and dependencies (#51)

General cleanup of dependencies and use statements across almost all
files.

Additionally several unused variables were eliminated.

timfelle committed:
Fix job submission and improve logging (#43)

This pull request includes several commits that fix issues related to
job submission and improve logging and messages to the user. The changes
ensure that the job submission works correctly on different clusters,
such as MareNostrum 5 and DTU. The submission process has been updated
to use batch scripts associated with the desired cluster, and the output
of the job is stored in the results folder after completion.
Additionally, the logging and messages to the user have been improved
for better clarity and understanding.

timfelle committed:
masks (#48)

Here I've added masks based on `point_zones` for both the design
indicator and source terms.
- volume constraint is now based on the optimization domain
- we have some basic masking functionality in `mask_ops` but I intend to
extend this in the future for more complex functionality (and GPUs)
- this is a big ***warning*** for Abbas (currently implementing the
`optimizer_t`) I've just masked the sensitivity. This is not the correct
way of handling the interface with MMA, because the $N$ is wrong. You
should use the masks on initialization of MMA such that only those
within the mask are flagged as design variables.

timfelle committed:
Enforce LF line endings and clean up code (#53)

This pull request enforces the use of LF line endings in the repository
and cleans up the code by removing unused variables and turning off a
warning that should not be an error. These changes improve code
consistency and maintainability.

timfelle committed:
Fix error log file name (#55)

The error log file name was changed from "error.err" to "error.log" in
order to align with naming conventions. This change ensures consistency
and improves clarity in the codebase.

timfelle committed:
Mesh modularity (#30)

Updates to the meshing script, we now support the following file
formats:

- `.jou`, journal files from Cubit
- `.geo`, gmesh geometry files
- `.exo`, Exodus files, not yet tested with multiblock files.
- `.re2`, just since we are often doing this as part of the others.

timfelle committed:
Update interface to new Neko interface (#57)

This pull request updates the interface to the new Neko interface.

timfelle committed:
Update documentation (#54)

The pull request includes several commits that update and improve the
documentation. The changes involve moving things back to the
documentation, cleaning up the documentation, and updating the table of
contents to render correctly. Additionally, the pull request ensures
that the documentation is not broken on pull requests and deploys the
documentation on the develop branch instead of the main branch.

timfelle committed:
Fix bug on how documentation is deployed.

timfelle committed:
Add missing device inclusions (#59)

timfelle committed:
Fix compatibility with neko (#62)

Remove unnecessary calls in the `neko_top_init` subroutine to improve
compatibility with neko.

timfelle committed:
Add JSON support for run files (#61)

Enable the run file to accept JSON files alongside existing formats,
enhancing flexibility in file handling. Update help messages and error
handling to reflect this new support.

timfelle committed:
Enhance run script robustness and usability (#63)

Improve the run script by adding a rerun option, skipping completed
runs, and enhancing status reporting. Clean up the functions script for
better organization and logging, ensuring more robust handling of
forceful stops.

timfelle committed:
Fix interfaces to Neko (#67)

timfelle committed:
fix interface to neko again (#68)

timfelle committed:
Update run script for progress tracking and cleanup logic (#65)

Refine the script to improve progress tracking by focusing on the
relevant log entries and adjust the cleanup condition for better
clarity.

Depends on: 

- [x] #67

timfelle committed:
Add immersed cylinder benchmark example (#66)

Introduce a benchmark case for the immersed boundary method featuring a
cylinder in a channel flow. Include necessary setup files, scripts, and
configurations to facilitate the simulation and data collection.

Depends on: 

- [x] #67

timfelle committed:
Prevent deletion of work folder during cleanup (#70)

Ensure the work folder is not deleted if the install directory is
contained within it during the cleanup process.

timfelle committed:
Update to new bc list interface (#69)

timfelle committed:
Feature/optimizer (#71)

- An optimizer type is added to improve the structure of the topology
optimization solver.
- MMA.f90 is updated to read the required parameters from the json case
file.
- Two different ways of scaling the objective and its derivatives is
considered.
- The rugby_ball example is updated to rugby_ball_optimizerobj to be use
the optimizer type.
- The old rugby_ball example is no longer supported due to changes in
MMA.f90.
to do list:
- The ...%free() functions needs to be checked and updated.

timfelle committed:
Enhance Brinkman flow examples and metrics computation (#64)

This update adds the examples for Brinkman flow, adds new metrics
computation features, and enhances the overall structure and robustness
of the code. It includes updates to the generator script, improved error
handling, and additional documentation for better usability.

Depends on: 

- [x] #67 

Closes #60

timfelle committed:
Handling of the problem type (#52)

This PR exposes some of the work we are doing in improving the problem
type.
The hope is to make it relatively generic and minimize which thing need
to be specified by a derived type.

This PR will primarily touch:

- How an optimization problem is defined (problem.f90, and derived
types)
- How the problem interact with a "design" or solution. (design.f90, and
detived types)
- Representation and handling of constraints and objectives.

timfelle committed:
Add verification example and revert previous handling changes (#73)

Introduce a verification example while reverting changes related to
problem type handling to restore previous functionality.

timfelle committed:
Add missing CUDA definition (#78)

Include a compile definition for CUDA support when the DEVICE_TYPE is
set to CUDA.

timfelle committed:
updated the jupyter notebook to what we had in the deliverable (#74)

I didn't realize this files was only on my computer locally.

timfelle committed:
Fix/bc_overhaul (#77)

timfelle committed:
Test working on a merge queue (#80)

A few updates to the CI. Some of this is to support the merge queues by
github.
The merge queue ensure multiple PR's being accepted at roughly the same
time does not break the repository. Each accepted Pr is added to a queue
and merged in order. This should help streamline our development, while
every merge is tested before the next is added.

- [x] Develop should properly support the merge queues.
- [x] The formatting action is now functional.
- [x] Compilation check should fail on warnings, such as un-initialized
variables.

timfelle committed:
Re introduction of the design class (#76)

The bugs have now ben fixed and this is a reintroduction of the updated
design types

Depends on #77

timfelle committed:
Hack support for the new solver updates (#83)

Enhance compatibility with the latest solver updates by modifying fluid
scheme references and ensuring proper initialization across various
subroutines.

timfelle committed:
Synchronize formatting action with Neko (#81)

Align the formatting process with Neko's standards and update the
contribution guidelines to reference Neko's documentation.

timfelle committed:
Fix MPI synchronization bug and clean up codebase (#82)

Address a bug that caused MPI de-synchronization and enhance code
clarity through multiple cleanup rounds. Additionally, update example
configurations and refine precision settings.

timfelle committed:
Fix buggy imports (#84)

Refactor module imports to ensure proper usage and avoid conflicts. This
improves code clarity and maintains consistency across modules.

timfelle committed:
Fix CMake compile definition and preprocessor directives for CUDA (#87)

Correct the application of the compile definition for CUDA in CMake and
update preprocessor directives to ensure proper conditional compilation.

timfelle committed:
Enhance Fortran build configuration and add LAPACK interface (#86)

Improve debugging options in the Fortran build configuration and
introduce a new LAPACK interface module for solving linear equations.
Additionally, refine type declarations for better clarity in the code.

timfelle committed:
Revert Fortran build configuration changes and improve debugging options (#89)

Revert previous enhancements to the Fortran build configuration while
relaxing debug options and addressing identified issues with array
temporaries.

timfelle committed:
MMA loop tweeks (#88)

Very small PR, but pretty important.

The design should be mapped to the brinkman term in between the MMA
update and running the simulation.

Also, stamping the zeroth iteration is nice for visualization.

timfelle committed:
Improve CMake structure and build process for examples (#90)

- Neko-TOP static library is built by default
- The `Examples` target now builds the examples.
- The setup script is updated to account for this.

Time to run `./setup.sh` (With all external dependencies ready)

- Before: 41 sec
- After: 26 sec

timfelle committed:
Improved logging (#91)

I guess we discussed this a bit this morning, but this is what I had in
mind.
I'm putting this in draft mode because I want comments before I move any
further with this:

- Do we think we'll want this in other optimization? ie, would a
`logger_t` be useful?
- Are we all happy with enforcing a log name to all
constraints/objectives so we can write them out?

But perhaps more importantly, what I've written currently is pretty
susceptible to truncation of that long string (buffer? is this the
correct word? that was the intention of the variable name `buff`)

### Example output

```
n =     86400, m =         1, tolerance =  1.00000000000000002E-03
iter, Total objective function, Dissipation, out of plane stresses, Volume constraint, KKTmax, KKTnorm2, scalaing factor
  0,     5.3696209634E+01,     4.0441625841E+01,     1.3254583793E+01,    -3.4113716238E-01,     1.7976931349+308,     1.7976931349+308,     1.0000000000E+00
  1,     5.3631210218E+01,     4.0441625841E+01,     1.3189584378E+01,    -3.4067827356E-01,     3.4063388244E-01,     1.8251180684E+00,     1.0000000000E+00
  2,     5.3266158113E+01,     4.0127814834E+01,     1.3138343279E+01,    -3.4021952588E-01,     3.4017508569E-01,     5.0791113981E+00,     1.0000000000E+00
  3,     5.2946050791E+01,     3.9894493099E+01,     1.3051557691E+01,    -3.3973026153E-01,     3.3968611728E-01,     4.1537413064E+00,     1.0000000000E+00
  4,     5.2488335978E+01,     3.9561169448E+01,     1.2927166530E+01,    -3.3922171690E-01,     3.3917795103E-01,     5.6123647507E+00,     1.0000000000E+00
  5,     5.1825567047E+01,     3.9082081599E+01,     1.2743485448E+01,    -3.3889046816E-01,     3.3884813432E-01,     7.9908112190E+00,     1.0000000000E+00
  6,     5.1000305086E+01,     3.8491289553E+01,     1.2509015532E+01,    -3.3852679847E-01,     3.3848677945E-01,     9.8497959414E+00,     1.0000000000E+00
  7,     5.0354702281E+01,     3.8025030555E+01,     1.2329671726E+01,    -3.3866646158E-01,     3.3862891930E-01,     7.8845237703E+00,     1.0000000000E+00
  8,     4.9785292383E+01,     3.7615237248E+01,     1.2170055135E+01,    -3.3902719413E-01,     3.3899308096E-01,     7.5623243319E+00,     1.0000000000E+00

```

### Example output using `csv_file_type`
```
iter, Total objective function, Dissipation, Out of plane stresses, Volume constraint, KKTmax, KKTnorm2, scalaing factor
0.0000000000000000,53.696209634065141,40.441625840817053,13.254583793248088,-0.34113716238378983,0.0000000000000000,0.0000000000000000,1.0000000000000000
1.0000000000000000,53.631210218451116,40.441625840817053,13.189584377634059,-0.34067827356064134,0.34063388243730330,1.8251180684119177,1.0000000000000000
2.0000000000000000,53.266158113095081,40.127814833617478,13.138343279477603,-0.34021952587850796,0.34017508568651300,5.0791113980976883,1.0000000000000000
3.0000000000000000,52.946050790610592,39.894493099386999,13.051557691223593,-0.33973026152689706,0.33968611727815884,4.1537413063900503,1.0000000000000000
4.0000000000000000,52.488335977948438,39.561169447724346,12.927166530224092,-0.33922171690237829,0.33917795102621462,5.6123647507375267,1.0000000000000000
5.0000000000000000,51.825567047220396,39.082081599088639,12.743485448131759,-0.33889046815908813,0.33884813432446081,7.9908112190203324,1.0000000000000000
6.0000000000000000,51.000305085635944,38.491289553207920,12.509015532428020,-0.33852679846645700,0.33848677945129146,9.8497959414055956,1.0000000000000000
7.0000000000000000,50.354702281281916,38.025030555367294,12.329671725914622,-0.33866646158438390,0.33862891930457578,7.8845237703490136,1.0000000000000000
8.0000000000000000,49.785292383215989,37.615237248149093,12.170055135066898,-0.33902719412641674,0.33899308096021136,7.5623243318792319,1.0000000000000000
```

timfelle committed:
Allow specification of processor count for execution (#92)

Introduce the ability to specify the number of processors to use via the
`NPROCS` environment variable, enhancing flexibility in resource
allocation during script execution.

timfelle committed:
Enhance JSON handling and driver structure improvements (#72)

Introduce json based initialization for all main types.

Additionally I have split the simulation out of the problem in order to
simplify dependencies.

Depends on #76

timfelle committed:
Gpu support (#79)

Originally this was 5 branches, and I wanted to submit 5 individual PRs
so this would be a bit cleaner.
But essentially this is bringing the Rugby ball case up to date with GPU
support.
The three branches were:
`GPU_support/masks`
`GPU_support/volume_constraint`
`GPU_support/objective_function`
`GPU_support/design` (which also contained the RAMP mappings etc)
`GPU_support/MMA_on_CPU` This one won't be needed, but it essentially
mem_copied everything from device to host, did MMA on the CPU, then
copied back.

But I was able to run the rugby ball on GPU so I was quite happy!
However, the optimization history wasn't EXACTLY the same as CPUs, so
I'm keeping this on draft until I can test more and resolve this.

**EDIT**

Ok I'm more and more confident the problem is with MMA on CPU, so I
think we can take this off draft mode. Also it looks like PR#85 needs a
lot of revisions and I want to get this PR through so we can all start
testing things on GPU. I apologies if there is bug is in here unrelated
to the hacky memcopies I performed.

 **note:**
I made a rather fundamental change to PDE filter, ie, solving for
$$\Delta \tilde{\rho}$$ not $$\tilde{\rho}$$, so already we won't match
with develop exactly, due to the residual prescribed by the PDE filter.

### Convergence history on develop
```
  0,  5.36962096340651414E+01, -3.41137162383789827E-01,  1.79769313486231571+308,  1.79769313486231571+308,  1.00000000000000000E+00
  1,  5.36962096340651414E+01, -3.40678273560641343E-01,  3.40633882437303304E-01,  1.82511436798967752E+00,  1.00000000000000000E+00
  2,  5.34909202977577962E+01, -3.40219526392907701E-01,  3.40175086197066656E-01,  2.93977673685759111E+00,  1.00000000000000000E+00
  3,  5.32662799775875442E+01, -3.39695799113361951E-01,  3.39651438883154300E-01,  3.19543034281953853E+00,  1.00000000000000000E+00
  4,  5.29593114254205517E+01, -3.39109389875194511E-01,  3.39065248562431587E-01,  3.98656352093350419E+00,  1.00000000000000000E+00
  5,  5.25307111679251335E+01, -3.38585236396372891E-01,  3.38541901793160127E-01,  5.22808835345366951E+00,  1.00000000000000000E+00
  6,  5.19234539710996330E+01, -3.38143391208646948E-01,  3.38102175627622070E-01,  7.03658279719623980E+00,  1.00000000000000000E+00
  7,  5.11597748945414921E+01, -3.37928412343896978E-01,  3.37888261694012759E-01,  8.82061398036337785E+00,  1.00000000000000000E+00
  8,  5.05602700893721249E+01, -3.38296226436742675E-01,  3.38259998372253234E-01,  6.98241437563093026E+00,  1.00000000000000000E+00
  9,  5.00465163087089451E+01, -3.38384508012592067E-01,  3.38349557059135442E-01,  6.54315441819684906E+00,  1.00000000000000000E+00
 10,  4.95253467948585779E+01, -3.38395758796641910E-01,  3.38362432950295389E-01,  6.75821664949387291E+00,  1.00000000000000000E+00
```

### Convergence history on CPU (with the new changes)
```
  0,  5.36962098917103532E+01, -3.41137162383789827E-01,  1.79769313486231571+308,  1.79769313486231571+308,  1.00000000000000000E+00
  1,  5.36962098917103532E+01, -3.40678272361158607E-01,  3.40633881225604174E-01,  1.82511372436923880E+00,  1.00000000000000000E+00
  2,  5.34909201511379564E+01, -3.40219524004039708E-01,  3.40175083796164401E-01,  2.93978525062977747E+00,  1.00000000000000000E+00
  3,  5.32662797551102756E+01, -3.39695795330679007E-01,  3.39651435093588583E-01,  3.19540660892341366E+00,  1.00000000000000000E+00
  4,  5.29593111639978673E+01, -3.39109383832420674E-01,  3.39065242501398190E-01,  3.98658874364438764E+00,  1.00000000000000000E+00
  5,  5.25307107108303555E+01, -3.38585227652321519E-01,  3.38541893041661768E-01,  5.22806768673648037E+00,  1.00000000000000000E+00
  6,  5.19234524546626943E+01, -3.38143379103141128E-01,  3.38102163503164688E-01,  7.03658086336280775E+00,  1.00000000000000000E+00
  7,  5.11597740019539202E+01, -3.37928397133592295E-01,  3.37888246463098230E-01,  8.82063926519529318E+00,  1.00000000000000000E+00
  8,  5.05602689953278457E+01, -3.38296209449853336E-01,  3.38259981376574370E-01,  6.98237874556235294E+00,  1.00000000000000000E+00
  9,  5.00465151924080729E+01, -3.38384488855007104E-01,  3.38349537884281071E-01,  6.54318127887768597E+00,  1.00000000000000000E+00
 10,  4.95253453940975277E+01, -3.38395744127610365E-01,  3.38362418271727805E-01,  6.75818420810124021E+00,  1.00000000000000000E+00
```

### Convergence history on GPU (with the new changes)
```
  0,  5.36962098917055499E+01, -3.41137162383794046E-01,  1.79769313486231571+308,  1.79769313486231571+308,  1.00000000000000000E+00
  1,  5.36962098917055499E+01, -3.40678272944622995E-01,  6.26706799280533833E+00,  6.43310832024656179E+02,  1.00000000000000000E+00
  2,  5.34910118155738559E+01, -3.40219525367835784E-01,  6.26705020196007023E+00,  6.43309822428294638E+02,  1.00000000000000000E+00
  3,  5.32664717580504288E+01, -3.39691202134478287E-01,  6.23507603889123274E+00,  6.41924745328890140E+02,  1.00000000000000000E+00
  4,  5.30975501242981096E+01, -3.39151984540635609E-01,  6.19951479473746137E+00,  6.40112443825842547E+02,  1.00000000000000000E+00
  5,  5.25350413808810828E+01, -3.38644965953474764E-01,  6.16274815428317435E+00,  6.38611032141696114E+02,  1.00000000000000000E+00
  6,  5.19248549714847698E+01, -3.38207041002288433E-01,  6.08546842737498661E+00,  6.34741181740812067E+02,  1.00000000000000000E+00
  7,  5.11622343791975069E+01, -3.37997668480812863E-01,  5.98907463234943904E+00,  6.30183729381854391E+02,  1.00000000000000000E+00
  8,  5.05718316461075759E+01, -3.38299594367291456E-01,  5.86614026930082400E+00,  6.24306904330888301E+02,  1.00000000000000000E+00
  9,  5.00919821259184630E+01, -3.38447606977853088E-01,  5.77873574707956283E+00,  6.19710815500791000E+02,  1.00000000000000000E+00
 10,  4.96736731441919943E+01, -3.38192325940090688E-01,  5.69855410966845621E+00,  6.15708752891213408E+02,  1.00000000000000000E+00
```

timfelle committed:
missing import in `mma_optimizer` (#96)

missing import

timfelle committed:
Fix verification run script and build examples from runner script (#94)

Streamline the execution process by building example executables from
the runner script.

Update the verification example. New filenames were not used properly.
Also the example will now check against the result folder, rather than a
manually added reference.

Also added an MPI guard on a few print statements

timfelle committed:
Bring rugby ball up to date (#97)

Here I've just updated the case file and driver of `rugby_ball` to
reflect the to updates in `rugby_verification`

**NOTE**

This will fail on device at the 2nd iteration due to the KKT check at
line 413 of `mma.f90`. Substituting
```fortran
    if (NEKO_BCKND_DEVICE .eq. 0) then
       call mma_KKT_cpu(this, x, df0dx, fval, dfdx)
    else
       write(stderr, *) "Device not supported for MMA."
       error stop
    end if
```

with
``` fortran
    if (this%backend == 'cpu') then
       call mma_KKT_cpu(this, x, df0dx, fval, dfdx)
    else
       write(stderr, *) "Device not supported for MMA."
       error stop
    end if
```

Is how PR#79 was run

timfelle committed:
Refactor/design-problem (#93)

This PR build on the structure.

The design, problem and simulation is untangled. The steady state
problem are no longer needed.

timfelle committed:
Documentation for adjoint fluid + name changes (#99)

The PR is moving in the direction of getting the adjoint passive scalar
off the ground. It consists of:
1) A bit of documentation for the adjoint velocity equations (I'll fill
in the adjoint passive scalar in a future PR). It isn't linked to the
rest of the docs, I'll wait for PR#95.
2) A lot of name changes such that `adjoint` becomes `adjoint_fluid`
where applicable, opening the way for us to have
- `fluid` and `adjoint_fluid`
- `scalar` and `adjoint_scalar`

timfelle committed:
Revert documentation changes and update adjoint references (#100)

Revert previous documentation changes related to adjoint fluid and
update references to use a unified adjoint scheme. Adjustments include
modifications to various modules and removal of outdated documentation.

timfelle committed:
this is what chat-GPT says to do :sweat_smile:  (#102)

what a kerfuffle...

according to chatGPT:

If you reverted the original PR using another PR, GitHub provides a
"Revert" button:

Go to the PR that reverted your original changes.
Click on the "Revert" button (if available). This will create a new PR
that restores the original changes.
Then, apply your fix on top of that before merging.

I hope that's what has been done in this PR, but please double check

timfelle committed:
Feature/gpu support mma (#85)

This updates the mma_optimizer structure, introduces mma_factory for CPU
and device backend types and generally, adds CUDA support for mma
solver.

timfelle committed:
Avoid built-in names and handle compilation errors (#104)

Ensure variable names do not conflict with built-in names and exit with
an error message if examples fail to compile.

timfelle committed:
really small change (#108)

I think the changes are obvious.
But I'd never seen that `"output_control": "never"` before. That's
really nice for us :+1:

gloopydoop committed:
Update for `neko` event (#111)

Update based on `neko` PR #1736

gloopydoop committed:
remove premultiply adjoint forcing with mass matrix (#112)

The pre-multiplication is handled in the source term handler, we
shouldn't have it here.

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Solve the mpi reduce issue in subsolve in mma_device. (#113)

Small fix in  mma_device.

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

timfelle committed:
Refactor the design_t (#103)

Major update to the design type.
This PR will improve the reuseability of the design type by hiding
implementation details behind a generic API.

timfelle committed:
Sync formatting to neko standards (#125)

Update formatting commands to align with neko's specifications.

Continuation lines starting with a `&` is not indented to 0 compared to
their current scope.

This is commonly used to break long strings.

gloopydoop committed:
Fix fluid ICs (#126)

This is in response to my `neko` PR passing the subdict to ICs.

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

timfelle committed:
Align with new KSP in neko (#129)

Update the usage of the Krylov solver to align with the new KSP
interface, replacing calls to `krylov_solver_destroy` with the
appropriate `free` method for memory management.

timfelle committed:
Cleanup and fix examples in the repository (#128)

Remove outdated and broken example files while fixing existing examples
to ensure proper functionality. This cleanup enhances the overall
quality and maintainability of the codebase.

timfelle committed:
Documentation improvements and example updates (#95)

Enhance documentation clarity and structure by updating references,
fixing formatting issues, and removing redundant content. Add missing
examples and address warnings in the documentation.

timfelle committed:
Align with new checkpoint structure in Neko (#133)

Update the code to align with the new checkpoint structure, ensuring
that time-related parameters are accessed correctly throughout the
adjoint and reset subroutines.

timfelle committed:
Ensure cross compilation support and clean up build flags (#122)

Introduce a new file for compiler flags to streamline the build process
and ensure compatibility with cross compilation. This change simplifies
the management of compiler options for different build types.

timfelle committed:
Fix aliasing module references (#123)

Update module references from `simulation` to `simulation_m` to resolve
aliasing issues. Adjust initialization and cleanup processes
accordingly.

Fixes: #116

gloopydoop committed:
Definition of steady residual + logging (#131)

I'm trying to dissect some of the changes that were originally in #110 .

Here I've changed the definition of the steady residual and added a
little csv writer so it can be monitored

gloopydoop committed:
Mapping cascade (#124)

This PR has a couple of nice features, but is still a bit of a work in
progress (maybe)

- The main contribution is the notion of a mapping cascade and a lot of
JSON support for the pre-existing mappings. Now we prescribe a list of
mappings which are performed sequentially, and it also takes care of the
chain rule moving backwards.
- There are some cosmetic changes to variable names, particularly
`sens_in` and `sens_out` :sweat_smile:
- Obviously much of the mapping was inspired by the source term handler,
so in principal this could be extended to allow user defined mappings in
some way, but this is to be discussed.

This PR is still in draft mode because I think we should really discuss
who owns the mapping, and how we should handle mapping to multiple
coeffients, and how we should hand constraints defined on intermediate
steps (ie, the potential for the volume constraint to defined on the
filtered design variable)

There are still a few cosmetics changes that need to be done so I'm
leaving this on draft mode, but take a look if you can and tell me what
you think

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
Optimization domain in JSON (#139)

Wow I can't believe this one was still hardcoded :sweat_smile: 
I've put the "help wanted" flag on because I don't know how this
translates into an optimization problem without a simulation. I guess
the word "domain" for topology optimization means a spatial domain, so
to me I think it should probably be rooted in
`optimization.design.domain`

But it's worth discussing.

Also aiming for many small PRs moving forward :sweat_smile:

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

timfelle committed:
Improve documentation and fix formatting issues (#138)

Enhance the documentation configuration and deployment, correct markdown
formatting, and complete doxygen comments across multiple files.
Addressed bugs in doxygen and made spelling updates for clarity.

timfelle committed:
Fix segmentation fault and improve logging (#134)

Address segmentation faults by ensuring proper memory management and
initialization. Enhance logging capabilities for better debugging and
add tests to verify functionality. Align code with new checkpoint
structures and clean up unnecessary outputs.

---------

Co-authored-by: Harrison Nobis <58177854+gloopydoop@users.noreply.github.com>

gloopydoop committed:
Initial material distribution (#127)

This one is pretty self explanatory.

But let me know your thoughts on moving the boundary_mesh utilities out
of the `brinkman_source_term` and into `set_scalar_ic` in `neko`, and
then we all have public access to that functionality.

I guess something also up for discussion is how multiple point zones is
handled. Currently it's quite nice in the `brinkman_source_term` but I
also saw that `point_zones` support boolean operations, so maybe this is
the way forward.

https://neko.cfd/docs/develop/da/dd0/point-zones.html

Depends on #124

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>
Co-authored-by: Abbas Mousavi <smousavi@kth.se>
Co-authored-by: Tim Felle Olsen <timfelle@hotmail.com>

timfelle committed:
Attempt to bring us up to speed with neko

timfelle committed:
progress

timfelle committed:
sync with neko

gloopydoop committed:
Passive scalar support (#110)

This PR contains:

-  Adjoint passive scalar support
- A scalar mixing objective function
- An example of a mixer

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>
Co-authored-by: Abbas Mousavi <smousavi@kth.se>
Co-authored-by: Tim Felle Olsen <timfelle@hotmail.com>

timfelle committed:
Merge remote-tracking branch 'origin/develop' into neko/pr_1727

gloopydoop committed:
Missing import (#143)

This was missing for compiling on device

timfelle committed:
Improve build process and device handling (#120)

Enhance the build process by automatically cleaning when the device type
changes, addressing issue #119. Adjust the NVCC wrapper to only set when
targeting CUDA, resolving issue #117. Update the JSON-Fortran linking to
use PkgConfig, fixing issue #114. This streamlines the setup and ensures
compatibility across different environments.

Fixes: #114
Fixes: #115 
Fixes: #117
Fixes: #119

---------

Co-authored-by: Abbas Mousavi <smousavi@kth.se>
Co-authored-by: gloopydoop <nobis@mech.kth.se>
Co-authored-by: Harrison Nobis <58177854+gloopydoop@users.noreply.github.com>

gloopydoop commented:
this looks good, it's still in draft mode and needs develop pulled in so I haven't tested it. But in principle it's approved :sweat_smile:

timfelle committed:
Add option to build documentation (#140)

Introduce a command-line option to build documentation during the
compilation process. This enhancement allows users to generate
documentation as part of the build workflow.

Depends on: #120

timfelle committed:
Merge branch 'develop' into neko/pr_1727

timfelle committed:
Merge branch 'develop' into neko/pr_1727

timfelle commented:
The PR is not yet in Neko's develop branch, which mean this is not yet needed. But it's here and ready to be merged

timfelle committed:
Merge commit 'b6eb97df5a4512377925ccfbaeb9ab7a850de83f' into cleanup

timfelle committed:
Merge branch 'cleanup' into neko/pr_1727

gloopydoop reviewed (approved):
looks good :+1:

timfelle committed:
Fix pFUnit
